### PR TITLE
Clean qr_solve update

### PR DESCRIPTION
**Pull request by scotty1024**
These changes merge what was a generic qr_solver out of a text book more firmly with Marlin's AUTO_BED_LEVELING_GRID function. The eixsting code was inlined and constants handed in from Marln_main.cpp referenced directly. Code path's that passed around flags to enable/disable bits of code were inspected and the path Marlin was using left and un-used code removed along with the bit inspection logic for enable/disable. Once the code was inlined and the actual Marlin code path exposed it was possible to see where the generic code was wrapping/copying arguments to be modular. This code was removed as well which allowed for runtime memory foot print reductions. All dynamic memory code was then easily removed as most of the wrapping/copying logic was gone (primary user of dynamic memory).

It was also seen that the code was calculating the value for TOL in a stand off generic fashion as well. It would scan the passed in A array of coordinates to find the maximum value (after making a copy of the array). The array is full of known values and thus it is possible to calculate the value for TOL without the need to scan the A array. In fact, due to the invariant nature of this array's contents: it may at some future point be possible to pre-factor it.

Observed a near 4KB reduction in Marlin code size!

All dynamic memory usage removed!

Runtime memory usage greatly reduced!

And a bug was found and removed!

### CODE REVIEW

**Comment by thinkyhead** [comment]
Looks good! If testers find no problems here I would like to see this merged as a bug fix, because this gets rid of the last vestiges of malloc / free in the code!

**Comment by AnHardt** [comment]
Somthing goes wrong here.

```
with this PR:                                                 wthout this PR:
N33 G28 *3                                                    N12 G28 *0
N36 G29 V4 *69                                                ok
ok                                                            N24 G29 V4 *70
G29 Auto Bed Leveling                                         G29 Auto Bed Leveling
Bed X: 10.000 Y: 51.000 Z: 1.099                              Bed X: 10.000 Y: 51.000 Z: 1.199
Bed X: 100.000 Y: 51.000 Z: 1.202                             Bed X: 100.000 Y: 51.000 Z: 1.283
Bed X: 190.000 Y: 51.000 Z: 1.305                             Bed X: 190.000 Y: 51.000 Z: 1.384
Bed X: 10.000 Y: 120.000 Z: 1.174                             Bed X: 10.000 Y: 120.000 Z: 1.253
Bed X: 100.000 Y: 120.000 Z: 1.226                            Bed X: 100.000 Y: 120.000 Z: 1.301
Bed X: 190.000 Y: 120.000 Z: 1.240                            Bed X: 190.000 Y: 120.000 Z: 1.311
Bed X: 10.000 Y: 189.000 Z: 1.161                             Bed X: 10.000 Y: 189.000 Z: 1.260
Bed X: 100.000 Y: 189.000 Z: 1.230                            Bed X: 100.000 Y: 189.000 Z: 1.310
Bed X: 190.000 Y: 189.000 Z: 1.152                            Bed X: 190.000 Y: 189.000 Z: 1.223
Eqn coefficients: a: 0.00348173 b: -0.00213868 d: 0.13533185  Eqn coefficients: a: 0.00038241 b: -0.00017814 d: 1.26371979
Mean of sampled points: 1.19883322                            Mean of sampled points: 1.28058326

Bed Height Topography:                                        Bed Height Topography:
+-----------+                                                 +-----------+
|...Back....|                                                 |...Back....|
|Left..Right|                                                 |Left..Right|
|...Front...|                                                 |...Front...|
+-----------+                                                 +-----------+
-0.03808 +0.03067 -0.04708                                    -0.02008 +0.02917 -0.05783
-0.02433 +0.02767 +0.04142                                    -0.02808 +0.02092 +0.03092
-1.66129 -2.17189 -1.04947                                    -0.08133 +0.00242 +0.10392
planeNormal x: -0.003482 y: 0.002139 z: 1.000000              planeNormal x: -0.000382 y: 0.000178 z: 1.000000
```

See the line before the last, in the first column.
The topography differs here considerably!
Are the values of the eqnBVector[] altered by the new qr-solve?
The resulting planeNormal also differs a lot.

Same bed. Z_PROBE_OFFSET_FROM_EXTRUDER differed by 0.1mm.

**Comment by AnHardt** [comment]
Made tests with 4x4 and 5x5 grid points. Always the first 3 elements of eqnBVector[] are altered.

File: Marlin/Marlin_main.cpp [hunk 1]
```diff
@@ -2574,8 +2574,18 @@ inline void gcode_G28() {
...
+	  /*
+	   * How much memory is consumed by changes in P?
+	   * (P*P*3*8) + (P*8) bytes, Thus a P of 10 uses 3200 bytes.
+	   *
+      	   * At present that would barely work on a PrintrBoard which is the
+	   * smallest board that can support ABL and only has 8K of RAM of which
+	   * about 4100 is already used up. 4100 + 3200 leaves a very tight
+	   * stack for the rest of this command to use.
+	   */
+          SERIAL_PROTOCOLPGM("?Number of probed (P)oints is implausible (2 minimum and 10 maximum).\n");
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp:2588]: Here you can just say `?Number of probed (P)oints is implausible (2-10).\n` since other errors use that convention.

**Comment by AnHardt** [comment]
An other example where the eqnBVector[] is corrupted: https://github.com/MarlinFirmware/Marlin/issues/2412#issuecomment-120650608

**Comment by scotty1024** [comment]
The new qr_solve uses the passed in A and B as residual work accumulators after it is done with their original contents. This cuts runtime memory usage almost in half.

There was a bug in the original qr_solve that affected its generated coefficient values. The new code should  be producing a different but more accurate result.

**Comment by avri-schneider** [comment]
@scotty1024 What is the hash of the commit with the _new_ `qr_solve()`?

I have applied the 3 commits:
scotty1024@c66cc56ba4258f705c5fbbb7257655e4b5d62df8
scotty1024@6dea4d89ee35d26542216eecf00f92dd7a5d69ef
scotty1024@ad61fd5efbb817a73fae602edf42cb6a101e4bc6

Yet had to revert them to get my auto bed leveling working.

Can you try running a test with a sloped bed, then issue:

```
G28
G29 V4
G1 Z0 F1000
G1 X15 Y15 F3000
G1 X150 Y150 F3000
G1 X150 Y15 F3000
G1 X15 Y150 F3000
```

And visually inspect that the distance of the nozzle from the bed remains constant at the 4 corners?

For me, distance of the nozzle from the bed remains constant without these 3 commits applied, but not constant with the 3 commits applied.

**Comment by AnHardt** [comment]
Now i faked in some calculated plane planes. I'm quite sure the new qr_solve has an error, because it gives not the same results as the 3 point equation.

1 front low - back high
2 left low - right high
3 front left low - back right high

**_grid old**_

```
 +0.50000 +0.50000 +0.50000
 +0.00000 +0.00000 +0.00000
 -0.50000 -0.50000 -0.50000
 planeNormal x: -0.005000 y: 0.000000 z: 1.000000
 +0.999987 +0.000000 +0.005000
 +0.000000 +1.000000 -0.000000
 -0.005000 +0.000000 +0.999987


 -0.50000 +0.00000 +0.50000
 -0.50000 +0.00000 +0.50000
 -0.50000 +0.00000 +0.50000
 planeNormal x: -0.000000 y: -0.005000 z: 1.000000
 +1.000000 +0.000000 +0.000000
 -0.000000 +0.999987 +0.005000
 -0.000000 -0.005000 +0.999987


 +0.00000 +0.25000 +0.50000
 -0.25000 +0.00000 +0.25000
 -0.50000 -0.25000 +0.00000
 planeNormal x: -0.002500 y: -0.002500 z: 1.000000
 +0.999997 +0.000000 +0.002500
 -0.000006 +0.999997 +0.002500
 -0.002500 -0.002500 +0.999994
```

**_3 point**_

```
 Bed X: 0.000 Y: 0.000 Z: 1.0
 Bed X: 0.000 Y: 200.000 Z: 1.0
 Bed X: 200.000 Y: 100.000 Z: 2.0
 +0.999987 +0.000000 +0.005000
 +0.000000 +1.000000 +0.000000
 -0.005000 +0.000000 +0.999987


 Bed X: 0.000 Y: 0.000 Z: 1.0
 Bed X: 0.000 Y: 200.000 Z: 2.0
 Bed X: 200.000 Y: 100.000 Z: 1.5
 +1.000000 +0.000000 +0.000000
 +0.000000 +0.999987 +0.005000
 +0.000000 -0.005000 +0.999987


 Bed X: 0.000 Y: 0.000 Z: 1.0
 Bed X: 0.000 Y: 200.000 Z: 1.5
 Bed X: 200.000 Y: 100.000 Z: 1.75
 +0.999997 +0.000000 +0.002500
 -0.000006 +0.999997 +0.002500
 -0.002500 -0.002500 +0.999994
```

**_grid new**_

```
 +0.50000 +0.50000 +0.50000
 +0.00000 +0.00000 +0.00000
 -1.50000 -2.46825 -1.46050 //-0.50000 -0.50000 -0.50000
 planeNormal x: 0.001974 y: -0.003026 z: 1.000000
 +0.999998 +0.000000 -0.001974
 +0.000006 +0.999995 +0.003026
 +0.001974 -0.003026 +0.999993


 -0.50000 +0.00000 +0.50000
 -0.50000 +0.00000 +0.50000
 -1.50000 -3.27512 -1.09825 //-0.50000 +0.00000 +0.50000
 planeNormal x: 0.004442 y: -0.004725 z: 1.000000
 +0.999990 +0.000000 -0.004442
 +0.000021 +0.999989 +0.004725
 +0.004442 -0.004725 +0.999979


 +0.00000 +0.25000 +0.50000
 -0.25000 +0.00000 +0.25000
 -1.50000 -2.87168 -1.27937 //-0.50000 -0.25000 +0.00000
 planeNormal x: 0.003208 y: -0.003876 z: 1.000000
 +0.999995 +0.000000 -0.003208
 +0.000012 +0.999993 +0.003875
 +0.003208 -0.003875 +0.999987
```

**Comment by scotty1024** [comment]
Very useful, thank you. I had been wondering if my using a Printrbot metal simple with MIC6 heated bed was an issue. I built it and leveled the bed carefully during construction. Yes, auto bed leveling is cool but running the Z axis up and down during the print doesn't speed it up. Seemed to me that the less Z moved the faster the prints would go.

So my bed is level to .07mm and the MIC6 is flat to essentially e from what I can tell.

Inside qr_solve the z heights get cooked down to a rank: kr. This process is where I found the bug, an m got passed where n should have been passed. The value was used in selecting the ranking

One thing I had noticed is that things work for some folks and seem not so good for others. After finding this bug I started to suspect the ranking selection was part of the happiness factor as the z height data from the bed drives the ranking as well eg tight bed vs looser bed.

I've been finishing a Prusa i3 the last week. It has a much looser bed. Not quite 1mm like yours but .7mm and the aluminum bed isn't MIC6 flat. I've built it with a Printrboard and a Printrbot sourced inductive probe (and a Printrbot sourced 12V inductive probe adaptor cable).

I've gotten X Y and Z steps setup and I'm presently struggling with probe calibration in X and Y. The probe is a massive 45mm away from the tip of the hot end and if it misses the bed the Z nuts get driven loose in the i3 design. It's been lots of fun resetting the z nuts so far. I'm getting very frustrated with why the code can't reliably handle a simple x and y offset for the probe. But rather seems drawn to dive the probe off the edge of the bed.

I also have a massive empathy with those talking about the issues in working with the development branch. :-(

One last thing I've discovered so far. Step calibration matters. I know duh right? We are using the probe to collect data and depending on the X Y and Z to move by the precise amount directed during the probing. Well I'm using a Printrboard that was not setup for a i3. So I got to witness how whacky the coefficients get when you tell it 100mm and the probe was more like 72mm and that .3mm z height reading was actually more like .7mm.

As I got my Brown and Sharpe out and started calibrating the auto bed leveling has been getting much more rational as well as I get the mm promised to actually be a mm delivered.

So a possible hint for those struggling with auto bed leveling: have you used a digital caliper to setup your actual steps per mm values? When I first started even my X and y didn't match each other much less z. Qr_solve is doing a _linear_ solution. If your X y and z aren't moving in linear amounts to each other: epic fail for auto bed leveling.

Yeah, sometimes it helps to eat all of the dog food. Yum yum!  :-)

**Comment by avri-schneider** [comment]
@scotty1024 enjoyed the read :+1: 

Are you saying your `qr_solve()` works and I just need to calibrate my `X`, `Y`, `Z` steps?
Why does it work with the old `qr_solve()` but not with yours?

**Comment by oysteinkrog** [comment]
@scotty1024 You should not have to calibrate X Y and Z, you calculate the step values for these axis directly based on your mechanics. 
See my config: https://github.com/oysteinkrog/Marlin/blob/config_i3_working/Marlin/Configuration.h

```
// default settings
// 1.8 degree NEMA17 motors
#define MOTOR_STEPS_PER_REV_AXIS_X 200
#define MOTOR_STEPS_PER_REV_AXIS_Y 200
#define MOTOR_STEPS_PER_REV_AXIS_Z 200
#define MOTOR_STEPS_PER_REV_AXIS_E 200

#define DRIVER_MICROSTEPS_AXIS_X 16
#define DRIVER_MICROSTEPS_AXIS_Y 16
#define DRIVER_MICROSTEPS_AXIS_Z 16
#define DRIVER_MICROSTEPS_AXIS_E 16

// T2 belts
#define BELT_PITCH_X 2.0
#define BELT_PITCH_Y 2.0

// 20-tooth pulleys
#define PULLEY_TEETH_X 20.0
#define PULLEY_TEETH_Y 20.0

// XY STEFPS
// steps_per_mm = (motor_steps_per_rev * driver_microstep) / (belt_pitch * pulley_number_of_teeth)
// NEMA 17 motor with T2 belt and 20-tooth pulley:
// (200 * 16) / (2 * 20) = 80.0
const float DEFAULT_X_STEPS = ((MOTOR_STEPS_PER_REV_AXIS_X*DRIVER_MICROSTEPS_AXIS_X) / (BELT_PITCH_X*PULLEY_TEETH_X));
const float DEFAULT_Y_STEPS = ((MOTOR_STEPS_PER_REV_AXIS_Y*DRIVER_MICROSTEPS_AXIS_Y) / (BELT_PITCH_X*PULLEY_TEETH_Y));

``
```

**Comment by thinkyhead** [comment]
I take it this still has some kinks to work out before it can be merged…. @AnHardt Are those first three elements of `eqnBVector` still getting stepped-on?

**Comment by AnHardt** [comment]
> Are those first three elements of eqnBVector still getting stepped-on?

Yes. @scotty1024 calls it reused.
But that is not the real problem. This qr_solve produces wrong rotation-matrices.
The test program, to prove this, is ready. Just waiting for an OK from @Roxy-3DPrintBoard to publish his code, what is included in the test.

**Comment by thinkyhead** [comment]
> Just waiting for an OK from @Roxy-3DPrintBoard to publish his code, what is included in the test.

His daddy was mean, naming him Roxanne. :wink:

**Comment by AnHardt** [comment]
Oh, sorry. I love Bryan Ferry and Roxy Music (https://www.youtube.com/watch?v=rpuiPhxsZLo) - guessed for a pseudonym. Women are so rare here - and at least 1/3 is male. :-)

**Comment by AnHardt** [comment]
Here a version of the test program without Roxy's code. https://github.com/AnHardt/qr_solve_test2

Even this shows that the current results from `qr_solve()` (this PR) are crap.
This also proves - the results from the old (now in use) `qr_solve()` are ok - at least for the tested cases.

**Comment by thinkyhead** [comment]
Hmm, maybe I will have a go at making the same general changes to `qr_solve` and see if I can avoid whatever pitfalls have led to the weird results…

**Comment by Roxy-3D** [comment]
> The test program, to prove this, is ready. Just waiting for an OK from @Roxy-3DPrintBoard to publish his code, what is included in the test.

@AnHardt    I missed your email asking permission to publish my code.   I finally saw it and responded with a "Go Ahead!"     Its OK to use whatever I do in whatever manner is helpful. It might be appropriate to put a #define in configuration.h to control which version of qr_solve gets compiled in.   The iterative version saves 10KB of code space.

**Comment by thinkyhead** [comment]
> control which version of `qr_solve` gets compiled

@AnHardt If the new version saves space and gives the same results, we should just drop the old one. Is this still exhibiting the same bugs, needing a fix? I still haven't dived into this issue.

**Comment by Roxy-3D** [comment]
> @AnHardt If the new version saves space and gives the same results, we should just drop the old one. Is this still exhibiting the same bugs, needing a fix? I still haven't dived into this issue.

I think we should go slow...   It is worth turning on the iterative solution and uncommenting the debug code at the very end of it.  You can watch it find the same solution that the real qr_solve() finds.   The problem is, we are using 4 byte floats and the precision is being pushed to the limit with all of the adding and subtracting.    I believe this code works.   I'm just saying...  Let's go slow...

**Comment by thinkyhead** [comment]
> I think we should go slow

I'm all for that with this one. Is there an idealized (pseudo-code) or Javascript version of this algorithm posted anywhere that I could play around with, I wonder…?

**Comment by Roxy-3D** [comment]
> I'm all for that with this one. Is there an idealized (pseudo-code) or Javascript version of this algorithm posted anywhere that I could play around with, I wonder…?

No...  This is sort of hacker code that was just done because it was fun and interesting.   But if you pull up the source code and look at it, it is super simple.   The debug code to print out the steps it is taking to solve the equation generates almost as much code as the code itself.   

I think we can pick up another 1/2 digit of precision if we put the D constant stuff at the the top of the loop instead of zero'ing in on A and B first.   But it actually works the way it is and we can play with the order that it solves the different constants later.

If you want to play with it, just load it into MSVC and give it a set of points to solve.   Or...  Turn on the debug code and watch it iterate in.   It is so simple, you are not going to need any help understanding it.

**Comment by scotty1024** [comment]
In case anyone is thinking I abandoned the code, I've just been waiting for any actionable comments.

Weird, broken (with no specifics), does it save memory, does it save FLASH...

As stated multiple times a transcription bug in the previous qr_solve was fixed, and location commented in the code of said fix. So weird and broken or just generates different and likely better due to the fix?

Does it save FLASH: compile it, I get a 4K reduction myself.

Does it save memory: yep, doesn't allocate several arrays the previous one used to protect A and B argument contents. This one uses the A and B as residual mathematical product holders. Hint: it isn't "corrupting" A and B through "pointer bugs", it's cleverly using them as the bigger sets of numbers get reduced to smaller and smaller sets of numbers ultimately reaching the final three numbers needed for the result.

As part of that the TOL value interacts with the factored Z heights to decide how many cubes  need to be considered to create the final plane (Kr factor). The bug I found affected this decision and people that don't have well calibrated printers will also affect this. This code is trying to perform a linear reduction and if your data isn't inside the probability domain usually referred to as a "bell curve" it will produce strange results. In essence TOL sets the size of the bell curve.

I don't know whom assembled this from the college text books but I'm pretty sure the current means of selecting TOL isn't well matched to this application (and is documented as just being a generic guess). The variant measurements susceptible to the e domain are in B but at present it is using the hard constants put into A to select TOL. These constant values are not governed by e so a poor value for TOL is thus being selected.

So I'm thinking TOL should be derived from the max(abs(B)) \* e. Having TOL based on the Measured Z data would I believe help with some of the frustration on loose printers that have much noise in their Z probing. I don't know that it would help bowed beds being run through it. The code can only handle so much non-linearity in the data before you start getting a rough whack result. As others have said: this code doesn't make your bed flat, it tries to make it level with the Z probe (which is itself hopefully level with the nozzle tip). In essence if the bed isn't flat enough it messes with the leveling.

I've seen this in action on two of my personal printers and the capture range for flatness noise isn't very large due to the present method of choosing TOL.

By the way, the code is butt ugly because it was originally written in Fortran and only lightly touched up when converting it to C. This and the maths involved make this block of code non-trivial to work with. I'm happy to assist in understanding the code if anyone is serious about learning what it is doing. But "it produces different numbers, it's broke" isn't inside my personal capture range.

**Comment by AnHardt** [comment]
Dear @scotty1024 
I admit, i have not looked into the code very deep, i just tested it.

For the most simple case, a perfectly level surface, i expect the correction matrix to be the unity matrix.

```
+1.000000 +0.000000 +0.000000 
+0.000000 +1.000000 +0.000000 
+0.000000 +0.000000 +1.000000 
```

With your code i get.

```
+0.999998 +0.000000 -0.001974 
+0.000006 +0.999995 +0.003026 
+0.001974 -0.003026 +0.999993 
```

Why?

**Comment by AnHardt** [comment]
Updated https://github.com/AnHardt/qr_solve_test2 to include @Roxy-3DPrintBoard s Iterative_LSF.
Add time measurement. You'll see, Iterative_LSF takes a good amount of time, but giving good results.

**Comment by AnHardt** [comment]
@thinkyhead 
Re-extracted qr_solve() from this PR today. No changes.

**Comment by Roxy-3D** [comment]
@AnHardt    I looked at that change you made to the line

```
distance = abs( A*eqn_A[i] + B*eqn_A[i+n] + C*eqn_B[i] + D) / normal ;
```

I don't think it is safe to remove the abs() function call.   The quantities A, B, and D get generated via additions and subtractions of delta to the base numbers.   I don't think we are guaranteed the sum of all those numbers is positive????   

Oh!  Wait...   C is 1.0 and doesn't change.   Are you thinking as long as the error from A, B, and D is less than 1.0 the number will always be positive?

**Comment by AnHardt** [comment]
@Roxy-3DPrintBoard 
The distance is not guaranteed to be > 0. But we are only interested in (distance \* distance). That is guaranteed to be > 0. (-*-=+)

**Comment by Roxy-3D** [comment]
@AnHardt 

> The distance is not guaranteed to be > 0. But we are only interested in (distance \* distance). That is guaranteed to be > 0. (-*-=+)

OK, but as a speed optimization, the (distance*distance) is not calculated.   Instead, the distance is divided by the sqrt(normal).   I think it is very easy to push the distance to a negative number if you have the bed slanted too much in the wrong direction (with a very small Z Probe Offset).

**Comment by AnHardt** [comment]
@Roxy-3DPrintBoard 
Are you not citing 

```
  for(int i = 0; i < n; i++) {
    distance = ( A*eqn_A[i] + B*eqn_A[i + n] + C*eqn_B[i] + D) / normal ;
    sum += distance*distance;
  }
  return sum;
```

?

**Comment by Roxy-3D** [comment]
You are correct!   Thank You for taking the time to point that out.

**Comment by AnHardt** [comment]
Made an attempt to make a pimped qr_solve(), with just the bare minimum changes, to remove malloc() and free(). Succeded! : -)
Updated the test program at: https://github.com/AnHardt/qr_solve_test2
A .diff with the changes is included because git sees `nm_qr_solve.cpp` as a new file.

An attempt to make a faster converging "Iterative_LSF" did not succeed. Tried to replace the "binary search" with a "Newton approximation", but lost too much precision.

**Comment by tonnico** [comment]
> Succeded! : -)

:+1: 
I will test this also on my printer (soon).
Does this also save memory?

**Comment by Roxy-3D** [comment]
> An attempt to make a faster converging "Iterative_LSF" did not succeed. Tried to replace the "binary search" with a "Newton approximation", but lost to much precision.

Did this attempt move the D constant stuff to the top of the loop?   My guess is moving the D constant stuff to the top of the loop will do two things.  It might add a 1/2 digit of precision but I suspect it will also converge quicker.

One more thought on speeding it up.   The summation_of_least_square_error_data() function does not need to normalize the data.   The line to calculate the normal variable and the division by it can be removed:

```
normal = sqrt( A*A + B*B + C*C);
for(i=0; i<n; i++) {
    distance = ( A*eqn_A[i] + B*eqn_A[i+n] + C*eqn_B[i] + D) / normal ;
    sum += distance*distance;
}
```

can be turned into:

```
for(i=0; i<n; i++) {
    distance = A*eqn_A[i] + B*eqn_A[i+n] + C*eqn_B[i] + D;
    sum += distance*distance;
}
```

**Pull request closed without merging**